src/rfc822.rs

```diff
@@ -59,32 +60,69 @@ impl<'s> Rfc822DateParser<'s> {
     #[inline]
     fn consume_u32(&mut self) -> Option<u32> {
```

Comment:
`consume_*` functions have been woefully inconsistent about if they should roll back in case of failure or not - either they all should or the callees should care about the error cases.

I suspect it's right for the consume function to remember where it was to begin with if they return a Result type. It's also probably a good idea to remove the push/pop position functions on the base parser because I think maintaining this information in the stack is probably nicer than having to handle this ourselves.